spec/GraphQLOverHTTP.md

```diff
@@ -112,72 +178,152 @@ http://example.com/product/graphql
 
-## Content Types
+A server MAY support requests encoded with and/or accepting other media types.
 
-The following are the officially recognized GraphQL content types to designate encoding JSON over HTTP.
+If the client does not supply a `Content-Type` header, the server SHOULD treat
+the request as if it had `Content-Type: application/json`.
 
-| Name  | Description |
-| ------------- | ------------- |
-| `application/graphql+json` | Required |
-| `application/json` | To support legacy clients |
+If the client does not supply an `Accept` header, the server SHOULD treat the
+request as if it had `Accept: application/json`.
 
-A server MUST support requests from clients with HTTP header `Content-Type: application/graphql+json` indicating that the body of the request is a JSON document with a GraphQL request. A server must indicate the content type of the response with a `Content-Type` header. This default value should be `Content-Type: application/graphql+json` indicating that the response is a valid JSON document [conforming to the GraphQL spec](https://spec.graphql.org/October2021/#sec-Response-Format).
+Note: These defaults are in place to maintain compatibility with legacy clients.
 
-A server MAY support requests from clients with other content types.
+Note: Prior to this specification, the media type `application/json` was in wide
+use for both the HTTP request body type and the HTTP response payload type.
+Unfortunately this causes a number of issues, not least that it means clients
+cannot trust responses from the server that do not use an HTTP 2xx status code
+(since these replies may come from non-compliant HTTP servers or proxies
+somewhere in the request chain). For this reason, this specification introduces
+the `application/graphql+json` media type; however, to give existing servers
+time to move over, it is not required to be supported until 1st January 2025.
 
-A client MUST handle receiving responses with HTTP header `Content-Type: application/graphql+json` since any compliant server must support this type.
+### Legacy watershed
 
-This header MAY include encoding information (e.g. `Content-Type: application/graphql+json; charset=utf-8`)
+From 1st January 2025 (`2025-01-01T00:00:00Z`), a server MUST support requests
+encoded with the `application/graphql+json` media type (as indicated by the
+`Content-Type` header).
+
+From 1st January 2025 (`2025-01-01T00:00:00Z`), a server MUST support requests
+which accept the `application/graphql+json` media type (as indicated by the
+`Accept` header).
 
 # Request
 
-A server MUST accept POST requests, and MAY accept other HTTP methods, such as GET.
+A server MUST accept POST requests, and MAY accept other HTTP methods, such as
+GET.
 
 ## Request Parameters
 
-A request for execution should contain the following request parameters:
+A {GraphQL-over-HTTP request} is formed of the following parameters:
+
+- {query} - A Document containing GraphQL Operations and Fragments to execute.
```

Comment:
```suggestion
- {query} - The string representation of a Document containing GraphQL Operations and Fragments to execute.
```

See https://github.com/graphql/graphql-over-http/issues/176 for context.

Reply:
I am choosing not to adopt this change for the reasons laid out in https://github.com/graphql/graphql-over-http/issues/176#issuecomment-1154019962 ; let's discuss it in a separate PR.

Reply:
How is interoperability possible if the client does not send a string?

Reply:
In GET, we state:

> The `query` parameter MUST be the string representation of the Source Text of
the Document as specified in [...]

In POST, we state:

> When encoded in JSON [...]
> - {query} - the string representation of the Source Text of the Document as specified in [...]

Thus for the standard encodings, string is specified for compatibility. However, alternative encodings are allowed by the spec (but not specified); these could come up with alternative serialization for the GraphQL query that is more efficient on the wire; for example you could consider a binary encoding that numbers all the keywords known to GraphQL and has an incredibly efficient encoding of a query; this should not fail the GraphQL-over-HTTP spec, it's something that we could later standardize even - e.g. `application/octet-graphql` may one day be included in this specification.

What matters for alternative (non-standard) encodings is what meaning the payloads must contain, not the specific stucture with which they do so.